add: adding loki details (#461)
microsoft/code-with-engineering-playbook: observability/tools/loki.md

# Loki

Loki is a horizontally-scalable, highly-available, multi-tenant log aggregation system, created by Grafana

> Labs inspired by the learnings from Prometheus. Loki is commonly referred as Prometheus but for logs, which

Reviewer: ```suggestion
Labs inspired by the learnings from Prometheus. Loki is commonly referred as 'Prometheus, but for logs', which
```

> makes total sense once both tools follow the same architecture, which is a agent collecting metrics in each

Reviewer: ```suggestion
makes total sense as both tools follow the same architecture, which is a agent collecting metrics in each
```

> makes total sense once both tools follow the same architecture, which is a agent collecting metrics in each

Reviewer: Suggestion to break the second part of the sentence into a separate sentence for clarity

Author: Just did some changes for better clarity, what do you think?

> makes total sense once both tools follow the same architecture, which is a agent collecting metrics in each

Reviewer 2: ```suggestion
which is a agent collecting metrics in each
```

```suggestion
 which is an agent collecting metrics in each
```

> makes total sense once both tools follow the same architecture, which is a agent collecting metrics in each

Reviewer 2: ```suggestion
makes total sense once both tools follow the same architecture, which is an agent collecting metrics in each
```

Author: nice touch, thanks!

of the components of the software system, a server which stores the logs and also the Grafana dashboard, which
access the loki server to build its visualizations and queries. That being said, Loki has three main
components:

## Promtail

> Its the agent portion of Loki. It can be used to grab metrics from several different places, like var/log/ for

Reviewer: ```suggestion
It is the agent portion of Loki. It can be used to grab metrics from several different places, like var/log/ for
```

> Its the agent portion of Loki. It can be used to grab metrics from several different places, like var/log/ for

Reviewer 2: Nit: promtail deals with log file vs metric sources (like Cortex/Prometheus does). I would replace the term `metrics` here with `logs`

There is a unified agent from Grafana which deals with both log files & metrics: https://grafana.com/docs/grafana-cloud/agent/

Author: That makes total sense as well!

> example. The configuration of the Promtail is a yaml file called ```config-promtail.yml``` commonly store on

Reviewer: ```suggestion
example. The configuration of the Promtail is a yaml file called ```config-promtail.yml``` commonly stored at
```

> the  ```/usr/local/bin``` path. In this file, its described all the paths and log sources that will be

Reviewer: Where is this file stored in other operating systems?

> the  ```/usr/local/bin``` path. In this file, its described all the paths and log sources that will be

Reviewer: the  ```/usr/local/bin``` path. In this file, its described all the paths and log sources that will be - From this sentence, it is not clear what is stored in this file. May be add a little bit more clarity around what is stored - file system paths of the log sources?

Author: I deleted the reference to the path! And made the sentence smaller. Does that help?

> the  ```/usr/local/bin``` path. In this file, its described all the paths and log sources that will be

Reviewer 2: Please double-check the pathing. `/usr/local/bin` is generally not where configuration files for system services go.

Author: Actually I just deleted reference to the path and it looks better! Thanks for calling that out!

aggregated on Loki Server.

## Loki Server

> Responsible for receiving and storing all the logs received from all the different systems. The Loki is also

Reviewer: ```suggestion
Loki Server is responsible for receiving and storing all the logs received from all the different systems. The Loki Server is also
```

> responsible for the queries done on Grafana, for example.

Reviewer: The queries from Grafana are processed on Loki Server?

Author: Yes! At least is what the doc says!

Reviewer 2: Yes, they are. I've contributed code to Loki in the past and understand its arch quite well.

## Grafana Dashboards

> 

Reviewer: It will be good to add that Grafana (>6.0) ships with Loki. Thoughts?

Author: Huuum, I think it makes sense! I am thinking about addning a note about it at the end of the Grafana section. Does that make sense?

Grafana Dashboards are responsible for creating the visualizations and performing queries. After all, it will
be a web page that people with the right access can log into to see, query and create alerts for the aggregated
logs.

## Why use Loki

The main reason to use Loki instead of other log aggregation tools, is that Loki optimizes the necessary
storage. It does that by following the same pattern as prometheus, which index the labels and make chunks
of the log itself, using less space than just storing the raw logs.

## References

> - [Loki Official Site](https://grafana.com/oss/loki/)

Reviewer: https://grafana.com/docs/loki/latest/best-practices/current-best-practices/

This is also a good link to add

Author: nice link, just added it!
